staging/src/k8s.io/apiserver/pkg/endpoints/filters/impersonation/constrained_impersonation_test.go

```diff
@@ -0,0 +1,822 @@
+		return authorizer.DecisionAllow, "", nil
+	}
+
+	if len(user.GetGroups()) > 0 && user.GetGroups()[0] == "scheduled-node-impersonater" && a.GetVerb() == "impersonate:associated-node" && a.GetResource() == "nodes" {
+		return authorizer.DecisionAllow, "", nil
+	}
+
+	if user.GetName() == "user-impersonater" && a.GetVerb() == "impersonate:user-info" && a.GetResource() == "users" {
+		return authorizer.DecisionAllow, "", nil
+	}
+
+	if len(user.GetGroups()) > 0 && user.GetGroups()[0] == "group-impersonater" && a.GetVerb() == "impersonate:user-info" && a.GetResource() == "groups" {
+		return authorizer.DecisionAllow, "", nil
+	}
+
+	if len(user.GetGroups()) > 0 && user.GetGroups()[0] == "extra-setter-scopes" && a.GetVerb() == "impersonate:user-info" && a.GetResource() == "userextras" && a.GetSubresource() == "scopes" {
+		return authorizer.DecisionAllow, "", nil
+	}
+
+	if user.GetName() == "legacy-impersonater" && a.GetVerb() == "impersonate" {
+		return authorizer.DecisionAllow, "", nil
+	}
+
+	if user.GetName() != "legacy-impersonator" &&
+		strings.HasPrefix(a.GetVerb(), "impersonate-on:") &&
+		(strings.HasSuffix(a.GetVerb(), "list") || strings.HasSuffix(a.GetVerb(), "get")) {
+		return authorizer.DecisionAllow, "", nil
+	}
+
+	return authorizer.DecisionNoOpinion, "deny by default", nil
+}
+
+func (c *constrainedImpersonateAuthorizer) attributeMatched(attrs []authorizer.AttributesRecord) error {
+	if len(attrs) != len(c.checkedAttrs) {
+		return fmt.Errorf("attributes not matched, get %d, expect %d", len(c.checkedAttrs), len(attrs))
+	}
+
+	for i := range c.checkedAttrs {
+		if c.checkedAttrs[i].GetAPIGroup() != attrs[i].GetAPIGroup() {
+			return fmt.Errorf("apigroup not matched, get %s, expect %s", c.checkedAttrs[i].GetAPIGroup(), attrs[i].GetAPIGroup())
+		}
+		if c.checkedAttrs[i].GetResource() != attrs[i].GetResource() {
+			return fmt.Errorf("resource not matched, get %s, expect %s", c.checkedAttrs[i].GetResource(), attrs[i].GetResource())
+		}
+		if c.checkedAttrs[i].GetVerb() != attrs[i].GetVerb() {
+			return fmt.Errorf("verb not matched, get %s, expect %s", c.checkedAttrs[i].GetVerb(), attrs[i].GetVerb())
+		}
+	}
+	return nil
+}
+
+func TestConstrainedImpersonationFilter(t *testing.T) {
+	testCases := []struct {
+		name                    string
+		user                    user.Info
+		impersonationUser       string
+		impersonationGroups     []string
+		impersonationUserExtras map[string][]string
+		impersonationUid        string
+		requestInfo             *request.RequestInfo
+		expectedAttributes      []authorizer.AttributesRecord
+		expectedUser            user.Info
+		expectedCode            int
+	}{
+		{
+			name: "impersonating-error",
+			user: &user.DefaultInfo{
+				Name: "tester",
+			},
+			expectedUser: &user.DefaultInfo{
+				Name: "tester",
+			},
+			expectedAttributes: []authorizer.AttributesRecord{
```

Comment:
You may want to create a helper to reduce the amount of boilerplate when creating an `AttributesRecord`.